### PR TITLE
Adds Plist Caching Feature

### PR DESCRIPTION
**Pull request by abigailshilts**
**Description**
Uses Plists to cache all conversation models that contain the current user
Will try to load from cache first when on the DisplayConversationsVC and updates the cache if new conversations have been created
Next Step is to implement this for DM models as well

**Milestone**
Milestone 7, subgoal 1

**Testing**
Added break points in code to ensure it wasn't at least right away running of of query results, also ran with wifi and cellular off: 

https://user-images.githubusercontent.com/58635711/181353962-d50bfe1e-8b91-4fdf-90ee-699df26b21ad.mp4

### CODE REVIEW

File: PickMeUp/PickMeUp/Caching/PMCachingFunctions.m [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
++(void)updateCache:(NSArray<PMConversation *> *)convos {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    NSError *error;
+    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
+    NSString *documentsDirectory = [paths objectAtIndex:0];
+
+    NSString *plistPath = [documentsDirectory stringByAppendingPathComponent:kConversationCacheFile];
+
+    if ([fileManager fileExistsAtPath:plistPath] == NO) {
+        NSString *resourcePath = [[NSBundle mainBundle] pathForResource:kConversationCache ofType:kPlistTitle];
```
Comment (nsazonov) [inline PickMeUp/PickMeUp/Caching/PMCachingFunctions.m:31]: What's in the default cache file?

File: PickMeUp/PickMeUp/Caching/PMCachingFunctions.m [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        [fileManager copyItemAtPath:resourcePath toPath:plistPath error:&error];
```
Comment (nsazonov) [inline PickMeUp/PickMeUp/Caching/PMCachingFunctions.m:32]: you should analyze error here, log it, and exit from the function

File: PickMeUp/PickMeUp/Caching/PMCachingFunctions.m [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+@interface PMCachingFunctions ()
+@end
+
+@implementation PMCachingFunctions
+
+static const NSString *const kConversationCacheFile = @"ConversationCache.plist";
+static const NSString *const kConversationCache = @"ConversationCache";
+
++(void)updateCache:(NSArray<PMConversation *> *)convos {
```
Comment (nsazonov) [inline PickMeUp/PickMeUp/Caching/PMCachingFunctions.m:22]: Would be nice to have a test coverage for these functions

File: PickMeUp/PickMeUp/ViewControllers/DM/PMDisplayConversationsViewController.m [hunk 4]
```diff
@@ -52,11 +58,15 @@ -(void)_runGetQuery {
...
-            for (PMConversation *convo in convos) {
-                [self.convoTree addConversation:convo];
+            if (convos.count > self.arrayToDisplay.count) {
+                for (int i = 0; i < convos.count-self.arrayToDisplay.count; i++) {
+                    [self.convoTree addConversation:convos[i]];
+                }
+                self.arrayToDisplay = convos;
+                [PMCachingFunctions updateCache:self.arrayToDisplay];
+                self.arrayToDisplay = [self.convoTree retreiveSubTree:kEmpt];
+                [self.tableView reloadData];
```
Comment (nsazonov) [inline PickMeUp/PickMeUp/ViewControllers/DM/PMDisplayConversationsViewController.m:68]: BTW, make sure this code runs on the main thread. It's not clear to me from the method name which thread is used to call the completion block.

**Review by nsazonov — APPROVED** [review]
Ok, this looks good. However, I recommend the following improvements.

1. Create a higher level object called something like PMDataService. This guy will use your cache under the hood. Your view controllers would work with `PMDataService` without knowing anything about caching. 
2. The API of `PMDataService` would probably looks like this
```
@property (nonatomic, strong) NSArray<PMConversation*> conversations // used by table view data source

- (void) fetchConversations:(Params)your params: (CompletionBlock)completion; // used by your view controller, which would reload table view on completion call
```
3. `fetchConversations` would first refer to cache and returned cached data right away, then it would make a network call and fetch new conversations, cache them and call completion block for the second time.
4. It's better to use background thread to save and retrieve data from cache. Use this API for that
```
dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
    //Background Thread
    dispatch_async(dispatch_get_main_queue(), ^(void){
        //Run UI Updates
    });
});
5. Add test case for your `PMDataService`
```

File: PickMeUp/PickMeUp/ViewControllers/DM/PMDisplayConversationsViewController.m [hunk 5]
```diff
@@ -52,11 +58,15 @@ -(void)_runGetQuery {
...
         if (convos != nil) {
-            for (PMConversation *convo in convos) {
-                [self.convoTree addConversation:convo];
+            if (convos.count > self.arrayToDisplay.count) {
+                for (int i = 0; i < convos.count-self.arrayToDisplay.count; i++) {
+                    [self.convoTree addConversation:convos[i]];
+                }
+                self.arrayToDisplay = convos;
+                [PMCachingFunctions updateCache:self.arrayToDisplay];
+                self.arrayToDisplay = [self.convoTree retreiveSubTree:kEmpt];
```
Comment (czamorag) [inline PickMeUp/PickMeUp/ViewControllers/DM/PMDisplayConversationsViewController.m:67]: nit: Typo in "retreiveSubTree"

File: PickMeUp/PickMeUp/Caching/PMCachingFunctions.m [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                                kSenderKey:[convo.sender fetchIfNeeded].username,
+                                kReceiverKey:[convo.receiver fetchIfNeeded].username,
+        };
+        [conversations addObject:toAdd];
+    }
+
+    [conversations writeToFile:plistPath atomically:YES];
+}
+
++(NSArray<PMConversation *> *)retreiveCache {
```
Comment (czamorag) [inline PickMeUp/PickMeUp/Caching/PMCachingFunctions.m:47]: nit: Typo in "retreive"

File: PickMeUp/PickMeUp/Caching/PMCachingFunctions.m [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+#import "PMCachingFunctions.h"
+#import "PMConversation.h"
+#import "StringsList.h"
+
+@interface PMCachingFunctions ()
+@end
+
+@implementation PMCachingFunctions
+
+static const NSString *const kConversationCacheFile = @"ConversationCache.plist";
```
Comment (czamorag) [inline PickMeUp/PickMeUp/Caching/PMCachingFunctions.m:19]: What's the reason to use a plist file instead of keychain or userdefaults?

Comment (nsazonov) [reply PickMeUp/PickMeUp/Caching/PMCachingFunctions.m:19]: I think file storage (plist) is a reasonable tool for this purpose. Keychain and NSUserDefaults are normally used to store  a smaller piece of data and access it by key later. CoreData uses file store to store the data.

**Pull request merged**